### PR TITLE
soundwire: bus: update Slave status in sdw_clear_slave_status

### PR DESCRIPTION
**Pull request by bardliao**
sdw_update_slave_status() will call the update_status() callback to
inform Slaves and do the corresponding tasks. We modify Slave status to
SDW_SLAVE_UNATTACHED but not update Slave status in sdw_clear_slave_status()
will lead to Slave status will never be updated elsewhere due to the
status is not changed.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

### CODE REVIEW

File: drivers/soundwire/bus.c [hunk 1]
```diff
@@ -1845,6 +1845,7 @@ void sdw_clear_slave_status(struct sdw_bus *bus, u32 request)
 		if (slave->status != SDW_SLAVE_UNATTACHED) {
```
Comment (kv2019i) [inline drivers/soundwire/bus.c:1845]: The commit message is a bit hard to read. How about: 
```
Call to sdw_update_slave_status() needs to be added to sdw_clear_slave_status()
to ensure Slaves are informed of the new status via update_status() callback.
```

Not sure that gets the whole message, but something along those lines is easier to understand. Basicly when setting a new set, modify() needs to be paired with update(), right?

**Review by kv2019i — CHANGES_REQUESTED** [review]
Patch is good, but the commit message could be improved a bit.

File: drivers/soundwire/bus.c (hunk 1 above)
Comment (bardliao) [reply drivers/soundwire/bus.c:1845]: Thanks @kv2019i for the commit message. Yes, modify() needs to be paired with update().

**Review by kv2019i — APPROVED** [review]
Thanks @bardliao !

**Comment by bardliao** [comment]
> Thanks @bardliao !

I would like to let @plbossart has a chance to review next week.

File: drivers/soundwire/bus.c (hunk 1 above)
Comment (ranj063) [reply drivers/soundwire/bus.c:1845]: if modify() needs to be paired with update(), why not call update() from within modify? Just wondering why only here and not in the other places where we call modify

Comment (bardliao) [reply drivers/soundwire/bus.c:1845]: We may need to do something between modify and update according to what status it is. We use [sdw_handle_slave_status](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/bus.c#L1715) to handle Slave status change event and we call modify and update within `sdw_handle_slave_status`. The other case is [sdw_clear_slave_status](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/bus.c#L1827), we use `sdw_clear_slave_status` to force Slave reattach in resume and somehow we didn't call `sdw_update_slave_status` within `sdw_clear_slave_status`

**Comment by plbossart** [comment]
I am not sure if this works in all cases @bardliao. The intent of the if (!hw_init) test in the resume() stage was to make sure we do not wait for the initialization complete in the case where there are 'ghost' devices exposed in ACPI tables but not physically populated in hardware. Are you sure this doesn't lead to reports of 'timeouts' with e.g. RT700 devices that are copy/pasted in all RVPs configs?

**Comment by plbossart** [comment]
> I am not sure if this works in all cases @bardliao. The intent of the if (!hw_init) test in the resume() stage was to make sure we do not wait for the initialization complete in the case where there are 'ghost' devices exposed in ACPI tables but not physically populated in hardware. Are you sure this doesn't lead to reports of 'timeouts' with e.g. RT700 devices that are copy/pasted in all RVPs configs?

I also don't know if the test for hw_init done *after* the wait_for_completion(initialization_complete) is actually doing anything. If the attach fails, then hw_init will not be set. Conversely if the device attaches, hw_init will be set already. Testing it in the resume doesn't accomplish anything.

Maybe the right fix is to use 

````c
if (!first_hw_init) // instead of if (!hw_init)
   return 0;

wait_for_completion(&initialization_complete);
````

That way if we resume a device that never attached then we just do nothing.

**Comment by bardliao** [comment]
> > I am not sure if this works in all cases @bardliao. The intent of the if (!hw_init) test in the resume() stage was to make sure we do not wait for the initialization complete in the case where there are 'ghost' devices exposed in ACPI tables but not physically populated in hardware. Are you sure this doesn't lead to reports of 'timeouts' with e.g. RT700 devices that are copy/pasted in all RVPs configs?
> 
> I also don't know if the test for hw_init done _after_ the wait_for_completion(initialization_complete) is actually doing anything. If the attach fails, then hw_init will not be set. Conversely if the device attaches, hw_init will be set already. Testing it in the resume doesn't accomplish anything.
> 
> Maybe the right fix is to use
> 
> ```c
> if (!first_hw_init) // instead of if (!hw_init)
>    return 0;
> 
> wait_for_completion(&initialization_complete);
> ```
> 
> That way if we resume a device that never attached then we just do nothing.

Ack

**Comment by plbossart** [comment]
closing since PR #2941 implements the same fixes.

**Pull request closed without merging**